### ISSUE TITLE
Deflate decompression doesn't fail for truncated input

### ISSUE THREAD

**Issue by vojtarylko** [apple/swift-nio-extras#175]
I tried to extend `HTTPRequestDecompressorTest` with this test case:

```swift
    func testDecompressionTruncatedInput() throws {
        // Truncated compressed data
        let compressed = ByteBuffer(bytes: [120, 156, 99, 0])

        let channel = EmbeddedChannel()
        try channel.pipeline.addHandler(NIOHTTPRequestDecompressor(limit: .none)).wait()
        let headers = HTTPHeaders([("Content-Encoding", "deflate"), ("Content-Length", "\(compressed.readableBytes)")])
        try channel.writeInbound(HTTPServerRequestPart.head(.init(version: .init(major: 1, minor: 1), method: .POST, uri: "https://nio.swift.org/test", headers: headers)))

        do {
            try channel.writeInbound(HTTPServerRequestPart.body(compressed))
            XCTFail("writeInbound should fail")
        } catch let error as NIOHTTPDecompression.DecompressionError {
            switch error {
            case .inflationError(Int(Z_BUF_ERROR)):
                // ok
                break
            default:
                XCTFail("Unexptected error: \(error)")
            }
        }
    }
```

As you can see I'm sending *truncated* invalid compressed data. I'd expect to get `DecompressionError.inflationError(-5)`. Instead decompression succeeds.

(I guess root cause is in `z_stream_s.inflatePart(to:minimumCapacity:)` which always calls `inflate` with `Z_NO_FLUSH` parameter.)

Is the problem in my test case, or in the NIO code?

**Comment by Lukasa** [comment]
`NIOHTTPRequestDecompressor` does not assume the message is finished until it receives a `HTTPServerRequestPart.end`. In this instance I suspect we _should_ see an error if we also send `.end`, but looking at the code I wonder if we're discarding the error result from dropping the decoder. It may be worth adding a send of `.end` after the send of `.body`, and then checking to see if there's an error reported here that we're losing: 

https://github.com/apple/swift-nio-extras/blob/5334d949febb396a4e2e5235e9fbcd9c3c014bb3/Sources/NIOHTTPCompression/HTTPDecompression.swift#L114

**Comment by vojtarylko** [comment]
The `inflateEnd(&self.stream)` returns just `Z_OK` when I add `try channel.writeInbound(HTTPServerRequestPart.end(nil))` into test.

**Comment by Lukasa** [comment]
Good catch, resolved by #177.
